src/sentry/utils/locking/lock.py

```diff
@@ -0,0 +1,34 @@
+import logging
+from contextlib import contextmanager
+
+
+logger = logging.getLogger(__name__)
+
+
+class Lock(object):
+    def __init__(self, backend, key, duration, routing_key=None):
+        self.backend = backend
+        self.key = key
+        self.duration = duration
+        self.routing_key = routing_key
+
+    def __repr__(self):
+        return '<Lock: {!r}>'.format(self.key)
+
+    def acquire(self):
+        self.backend.acquire(self.key, self.duration, self.routing_key)
+
+        @contextmanager
+        def releaser():
+            try:
+                yield
+            finally:
+                self.release()
+
+        return releaser()
+
+    def release(self):
+        try:
+            self.backend.release(self.key, self.routing_key)
+        except Exception as error:
+            logger.warning('Failed to release %r due to error: %r', self, error, exc_info=True)
```

Comment:
I couldn't think of a case (that we have, at least) where losing the lock would allow us to actually take any corrective action. The situations I can think of that would cause this:

- Redis crashes, so you're unable to communicate to the database at all to release the lock (so nobody else is grabbing it anyway),
- you already obtained and released the lock and are trying to release it again,
- you overrun the duration of your lease on the lock and can't release a lock that another process already obtained,
- you overrun the duration of your lease on the lock so it's expired and you can't delete the key (or be sure that another process didn't obtain and release it in the meantime.)

I'm not sure any of these situations necessitate an exception (since any potential damage has already been done) but it does seem like something that we'd want to know about, which is why I just logged it as a warning. I guess in summary, I'm not really *opposed* to making it an exception, I'm just not sure if there is any benefit to doing that either.